reference/src/glossary.md

```diff
@@ -1,5 +1,57 @@
 ## Glossary
 
+#### Aliasing
+
+(Please note: a full aliasing model for Rust has not yet been constructed, but
+at the moment we can give the following definition. The most developed potential
+aliasing model so far is known as "Stacked Borrows", and can be found
+[here](https://github.com/Lokathor/unsafe-code-guidelines/blob/lokathor/wip/stacked-borrows.md).)
+
+*Aliasing* is any time one pointer or reference points to a "span" of memory
+that overlaps with the span of another pointer or reference. A span of memory is
+similar to how a slice works: there's a base byte address as well as a length in
+bytes.
+
+Consider the following example:
+
+```rust
+fn main() {
+    let u: u64 = 7_u64;
+    let r: &u64 = &u;
+    let s: &[u8] = unsafe {
+        core::slice::from_raw_parts(&u as *const u64 as *const u8, 8)
+    };
+    let (head, tail) = s.split_first().unwrap();
+}
+```
+
+In this case, both `r` and `s` alias each other, since they both point to all of
+the bytes of `u`.
+
+However, `head` and `tail` do not alias each other: `head` points to the first
+byte of `u` and `tail` points to the other seven bytes of `u` after it. Both `head`
+and `tail` alias `s`, any overlap is sufficient to count as an alias.
+
+* The span of a pointer or reference is the size of the value being pointed to or referenced.
+* For some type `T` that is [`Sized`](https://doc.rust-lang.org/core/marker/trait.Sized.html)
+  The span length of a pointer or reference to `T` is found with `size_of::<T>()`.
+* When `T` is not `Sized` the story is a little tricker:
+  * If you have a reference `r` you can use `size_of_val(r)` to determine the
+    span of the reference.
+  * If you have a pointer `p` you must unsafely convert that to a reference before
+    you can use `size_of_val`. There is not currently a safe way to determine the
+    span of a pointer to an unsized type.
+* The Data Layout chapter also has more information on the sizes of different types.
```

Comment:
What is the purpose of this note? It surely shouldn't be part of the *definition*.

Reply:
Well, it is impossible to safely use `size_of_val` with `*const [T]` and `*mut [T]` if you don't know the span size that must be valid and not aliasing but you _can_, if you're feeling spicy, transmute your fat pointer into a layout-equivalent `DynObject` or `Slice<T>` and then read the `usize` field, for example.

It's worth a mention that people should go check that chapter I think.

Reply:
But that's covered already by "unsafely convert that to a reference".

Mentioning going to the chapter seems fine (then it should be a link though). But it's not part of the definition of a "span", so it should be *after* the enumeration, not *in* the enumeration.

Reply:
Okay, between this point and the other unresolved point I now understand that you're looking at this list _very_ differently from how I intended it. But we can go with that.